File: pipeline/download.py
"""Checksummed, resumable, retrying downloads into pipeline/.cache.

Every source file is pinned by SHA-256. Where the publisher also gives an MD5 (USGS does),
it is checked too, so the first download of a new source is verified against the
publisher rather than only against itself. A download that does not match is discarded and
retried from scratch; it is never used. Cached files are re-verified on every run.
"""

from __future__ import annotations

import hashlib
import time
import urllib.request
from pathlib import Path

CACHE = Path(__file__).resolve().parent / ".cache"
ATTEMPTS = 6
CHUNK = 1 << 20
# Identify ourselves. planetarymaps.usgs.gov answers 403 to Python's default "Python-urllib".
USER_AGENT = "solar-system-pipeline/0 (+https://github.com/asrijanga/solar-system)"


def digest(path: Path, algorithm: str) -> str:
    h = hashlib.new(algorithm)
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(CHUNK), b""):
            h.update(chunk)
    return h.hexdigest()


def sha256(path: Path) -> str:
    return digest(path, "sha256")


def _verified(path: Path, sha: str | None, md5: str | None) -> tuple[bool, str]:
    if md5 is not None:
        actual = digest(path, "md5")
        if actual != md5:
            return False, f"md5 mismatch: got {actual}, publisher says {md5}"
    if sha is not None:
        actual = sha256(path)
        if actual != sha:
            return False, f"sha256 mismatch: got {actual}"
    return True, ""


def fetch(url: str, expected_sha256: str | None, name: str, *, md5: str | None = None) -> Path:
    """Returns a verified local copy of `url`, resuming a partial download if one exists.

    `expected_sha256` may be None only for a first download that has a publisher MD5; the
    caller must then pin the SHA-256 this prints.
    """
    if expected_sha256 is None and md5 is None:
        raise ValueError(f"{name}: refusing an unpinned download with no publisher checksum")
    CACHE.mkdir(exist_ok=True)
    target = CACHE / name
    if target.exists():
        ok, _ = _verified(target, expected_sha256, md5)
        if ok:
            return target
        target.unlink()

    partial = target.with_suffix(target.suffix + ".partial")
    last_error = "no attempt made"
    for attempt in range(1, ATTEMPTS + 1):
        try:
            have = partial.stat().st_size if partial.exists() else 0
            headers = {"User-Agent": USER_AGENT}
            if have:
                headers["Range"] = f"bytes={have}-"
            request = urllib.request.Request(url, headers=headers)
            with urllib.request.urlopen(request, timeout=120) as response:
                resumed = have > 0 and response.status == 206
                length = response.headers.get("Content-Length")
                total = (have if resumed else 0) + int(length) if length is not None else None
                with partial.open("ab" if resumed else "wb") as out:
                    while chunk := response.read(CHUNK):
                        out.write(chunk)
            size = partial.stat().st_size
            if total is not None and size < total:
                # The connection ended early without an error: keep what arrived and resume.
                raise OSError(f"transfer ended early at {size} of {total} bytes")
            ok, why = _verified(partial, expected_sha256, md5)
            if ok:
                partial.replace(target)
                if expected_sha256 is None:
                    print(f"  {name}: verified against publisher md5; pin sha256 {sha256(target)}")
                return target
            last_error = why
            partial.unlink()  # a complete but wrong file: start again
        except OSError as error:  # network errors, truncated transfers: resume next time
            last_error = str(error)
        print(f"  download attempt {attempt} of {url} failed: {last_error}", flush=True)
        time.sleep(min(2**attempt, 30))
    raise RuntimeError(f"could not download {url}: {last_error}")

File: tools/data/download.ts
// Checksummed, resumable, retrying downloads into pipeline/.cache: the TypeScript twin of
// pipeline/download.py, sharing its cache so a file fetched by either is reused by both.
//
// Every source is pinned by SHA-256. A download that does not match is discarded and fetched
// again; it is never used. Cached files are re-verified on every run.

import { createHash } from 'node:crypto';
import {
  createReadStream,
  createWriteStream,
  existsSync,
  mkdirSync,
  renameSync,
  statSync,
  unlinkSync,
} from 'node:fs';
import { join } from 'node:path';
import { Readable } from 'node:stream';
import { pipeline } from 'node:stream/promises';

export const CACHE = join(import.meta.dirname, '..', '..', 'pipeline', '.cache');
const ATTEMPTS = 6;
const USER_AGENT = 'solar-system-pipeline/0 (+https://github.com/asrijanga/solar-system)';

export async function sha256(path: string): Promise<string> {
  const hash = createHash('sha256');
  await pipeline(createReadStream(path, { highWaterMark: 1 << 20 }), hash);
  return hash.digest('hex');
}

export interface FetchOptions {
  /** Inclusive start, exclusive end: fetch only these bytes, by HTTP range request. */
  readonly byteRange?: readonly [number, number];
}

/** A verified local copy of `url` (or of a byte range of it), as `name` in the cache. */
export async function fetchPinned(
  url: string,
  expectedSha256: string,
  name: string,
  options: FetchOptions = {},
): Promise<string> {
  mkdirSync(CACHE, { recursive: true });
  const target = join(CACHE, name);
  if (existsSync(target)) {
    if ((await sha256(target)) === expectedSha256) return target;
    unlinkSync(target);
  }
  const partial = `${target}.partial`;
  let lastError = 'no attempt made';
  for (let attempt = 1; attempt <= ATTEMPTS; attempt++) {
    try {
      const have = existsSync(partial) ? statSync(partial).size : 0;
      const headers: Record<string, string> = { 'User-Agent': USER_AGENT };
      if (options.byteRange !== undefined) {
        const [start, end] = options.byteRange;
        headers['Range'] = `bytes=${start + have}-${end - 1}`;
      } else if (have > 0) {
        headers['Range'] = `bytes=${have}-`;
      }
      const response = await fetch(url, { headers });
      if (options.byteRange !== undefined && response.status !== 206) {
        throw new Error(`server ignored the range request (HTTP ${response.status})`);
      }
      if (!response.ok || response.body === null) throw new Error(`HTTP ${response.status}`);
      const resumed = have > 0 && response.status === 206;
      const length = response.headers.get('content-length');
      const total = length === null ? null : (resumed ? have : 0) + Number(length);
      await pipeline(
        Readable.fromWeb(response.body as never),
        createWriteStream(partial, { flags: resumed ? 'a' : 'w' }),
      );
      const size = statSync(partial).size;
      // The connection ended early without an error: keep what arrived and resume.
      if (total !== null && size < total)
        throw new Error(`transfer ended early at ${size} of ${total} bytes`);
      const actual = await sha256(partial);
      if (actual === expectedSha256) {
        renameSync(partial, target);
        return target;
      }
      lastError = `sha256 mismatch: got ${actual}`;
      unlinkSync(partial); // a complete but wrong file: start again
    } catch (error) {
      lastError = String(error);
    }
    console.log(`  download attempt ${attempt} of ${url} failed: ${lastError}`);
    await new Promise((resolve) => setTimeout(resolve, Math.min(2 ** attempt, 30) * 1000));
  }
  throw new Error(`could not download ${url}: ${lastError}`);
}
